### PR TITLE
Testing of edg

### PR DESCRIPTION
**Pull request by GeneralDenmark**
This small pr. samples some smaller and a large test to be run @ the edg.

### CODE REVIEW

**Comment by MathiasMehl** [comment]
test_small_dg_all_true_except_for_d(1)
test_small_dg_all_true_except_for_c(1)
test_with_edg(1)
test_loop_di_loops(1)
test_negation_edges(1)

**Review by NicEastvillage — APPROVED** [review]
Nice. Bugs confirmed.

**Review by MathiasMehl — APPROVED** [review]

File: src/edg.rs [hunk 1]
```diff
@@ -424,3 +425,436 @@ impl<B: Broker<V> + Debug, G: ExtendedDependencyGraph<V> + Send + Sync + Debug,
...
+        }
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::A, 1), VertexAssignment::TRUE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::B, 1), VertexAssignment::TRUE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::C, 1), VertexAssignment::TRUE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::D, 1), VertexAssignment::TRUE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::E, 1), VertexAssignment::TRUE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::F, 1), VertexAssignment::TRUE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::T, 1), VertexAssignment::FALSE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::G, 1), VertexAssignment::FALSE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::N, 1), VertexAssignment::FALSE);
```
Comment (jenrik) [inline src/edg.rs:592]:
```suggestion
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::A, 1), VertexAssignment::TRUE, "Vertex A");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::B, 1), VertexAssignment::TRUE, "Vertex B");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::C, 1), VertexAssignment::TRUE, "Vertex C");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::D, 1), VertexAssignment::TRUE, "Vertex D");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::E, 1), VertexAssignment::TRUE, "Vertex E");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::F, 1), VertexAssignment::TRUE, "Vertex F");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::T, 1), VertexAssignment::FALSE, "Vertex T");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::G, 1), VertexAssignment::FALSE, "Vertex G");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::N, 1), VertexAssignment::FALSE, "Vertex N");
```

File: src/edg.rs [hunk 2]
```diff
@@ -424,3 +425,436 @@ impl<B: Broker<V> + Debug, G: ExtendedDependencyGraph<V> + Send + Sync + Debug,
...
+
+                        successors
+                    }
+                }
+            }
+        }
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::A, 1), VertexAssignment::TRUE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::B, 1), VertexAssignment::TRUE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::C, 1), VertexAssignment::FALSE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::D, 1), VertexAssignment::TRUE);
```
Comment (jenrik) [inline src/edg.rs:665]:
```suggestion
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::A, 1), VertexAssignment::TRUE, "Vertex A");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::B, 1), VertexAssignment::TRUE, "Vertex B");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::C, 1), VertexAssignment::FALSE, "Vertex C");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::D, 1), VertexAssignment::TRUE, "Vertex D");
```

File: src/edg.rs [hunk 3]
```diff
@@ -424,3 +425,436 @@ impl<B: Broker<V> + Debug, G: ExtendedDependencyGraph<V> + Send + Sync + Debug,
...
+
+                        successors
+                    }
+                }
+            }
+        }
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::A, 1), VertexAssignment::FALSE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::B, 1), VertexAssignment::FALSE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::C, 1), VertexAssignment::FALSE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::D, 1), VertexAssignment::TRUE);
```
Comment (jenrik) [inline src/edg.rs:738]:
```suggestion
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::A, 1), VertexAssignment::FALSE, "Vertex A");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::B, 1), VertexAssignment::FALSE, "Vertex B");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::C, 1), VertexAssignment::FALSE, "Vertex C");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::D, 1), VertexAssignment::TRUE, "Vertex D");
```

File: src/edg.rs [hunk 4]
```diff
@@ -424,3 +425,436 @@ impl<B: Broker<V> + Debug, G: ExtendedDependencyGraph<V> + Send + Sync + Debug,
...
+
+                        successors
+                    }
+                }
+            }
+        }
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::A, 1), VertexAssignment::FALSE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::B, 1), VertexAssignment::FALSE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::C, 1), VertexAssignment::TRUE);
+        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::D, 1), VertexAssignment::TRUE);
```
Comment (jenrik) [inline src/edg.rs:811]:
```suggestion
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::A, 1), VertexAssignment::FALSE, "Vertex A");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::B, 1), VertexAssignment::FALSE, "Vertex B");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::C, 1), VertexAssignment::TRUE, "Vertex C");
        assert_eq!(distributed_certain_zero(ExampleEDG {}, ExampleEDGVertices::D, 1), VertexAssignment::TRUE, "Vertex D");
```

**Review by jenrik — CHANGES_REQUESTED** [review]
It would be nice to have the vertex names attached to each assertion like this `assert_eq!(distributed_certain_zero(...), Vertices::A, "Vertex A")` so you can see which test case failed clearly and obviously.

sidenote, I find the use of inline declarations of enums and impls in functions to be brilliant. I didn't know that was possible 👍 👍 👍

**Comment by GeneralDenmark** [comment]
@jenrik yes, your absolutely right, I remember that I though it was annoying that I didn't know which assert_eq that failed when I wrote the tests. Yet, I kinda forgot it, and just went with looking @ line numbers instead :sweat_smile:

**Review by jenrik — APPROVED** [review]

**Pull request merged**